pennylane/grouping/utils.py

```diff
@@ -0,0 +1,603 @@
+    array([1., 1., 0., 0., 1., 1.])
+    >>> pauli_to_binary(qml.PauliX('c') @ qml.PauliY('a') @ qml.PauliZ('b'), wire_map=wire_map)
+    array([1., 0., 1., 1., 1., 0.])
+
+    Now the two Pauli words are distinct in the binary vector representation, as the vector
+    components are consistently mapped from the wire labels, rather than enumerated
+    left-to-right.
+
+    If `n_qubits` is unspecified, the dimensionality of the vector representation will be inferred
+    from the size of support of the Pauli word,
+
+    >>> pauli_to_binary(qml.PauliX(0) @ qml.PauliX(1))
+    array([1., 1., 0., 0.])
+    >>> pauli_to_binary(qml.PauliX(0) @ qml.PauliX(5))
+    array([1., 1., 0., 0.])
+
+    Dimensionality higher than twice the support can be specified by `n_qubits`,
+
+    >>> pauli_to_binary(qml.PauliX(0) @ qml.PauliX(1), n_qubits=6)
+    array([1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
+    >>> pauli_to_binary(qml.PauliX(0) @ qml.PauliX(5), n_qubits=6)
+    array([1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
+
+    For these Pauli words to have a consistent mapping to vector representation, we once again
+    need to specify a `wire_map`.
+
+    >>>wire_map = {Wires(0):0, Wires(1):1, Wires(5):5}
+    >>> pauli_to_binary(qml.PauliX(0) @ qml.PauliX(1), n_qubits=6, wire_map=wire_map)
+    array([1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
+    >>> pauli_to_binary(qml.PauliX(0) @ qml.PauliX(5), n_qubits=6, wire_map=wire_map)
+    array([1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.])
+
+    Note that if `n_qubits` is specified and `wire_map` is specified, the dimensionality of the
+    vector representation will be inferred from the highest integer in `wire_map.values()`.
+
+    >>>wire_map = {Wires(0):0, Wires(1):1, Wires(5):5}
+    >>> pauli_to_binary(qml.PauliX(0) @ qml.PauliX(5),  wire_map=wire_map)
+    array([1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.])
+
+    Args:
+        pauli_word (Union[Identity, PauliX, PauliY, PauliZ, Tensor]): the Pauli word to be
+            converted to binary vector representation.
+
+    Keyword args:
+        n_qubits (int): number of qubits to specify dimension of binary vector representation.
+        wire_map (dict): dictionary containing all wire labels used in the Pauli word as keys, and
+                         unique integer labels as their values
+
+    Returns:
+        array: the 2*n_qubits dimensional binary vector representation of the input Pauli word.
+
+    Raises:
+        TypeError: if the input `pauli_word` is not an instance of Identity, PauliX, PauliY,
+            PauliZ or tensor products thereof.
+        ValueError: if `n_qubits` is less than the number of wires acted on by the Pauli word.
+    """
+
+    if not is_pauli_word(pauli_word):
+        raise TypeError(
+            "Expected a Pauli word Observable instance, instead got {}.".format(pauli_word)
+        )
+
+    if isinstance(pauli_word, Tensor):
+        pauli_word = pauli_word.prune()
```

Comment:
I had originally made `pauli_to_binary` infer the output vector dimensionality from the _non-trivial_ (not including action 
 of identities) support of the input Pauli word, if `n_qubits` is unspecified. Since the non-trivial support of tensor products of identity is the empty set, the output vector dimensionality is lowest it can be, i.e. dim=2. Dimensionality is inferred from the non-trivial support since implicitly there always exist single-qubit identity operations (e.g. X1 Y2 = X1 Y2 I3 I4...) and `pauli_to_binary` does not differentiate if the `Identity` instances are actually included in the `Tensor` instance or not. I don't see a problem with changing this so that `Identity` instances are included when obtaining the support of the Pauli word (so your example would give a vector of dim=2*3). One just needs to be more careful about including `Identity` instances in their Pauli words when they aren't needed, as they risk using a binary vector representation larger than what's necessary.

Reply:
I suppose it comes down to whether or not there is some utility in including `Identity` instances in an `Observable` instance. If so, I can make the appropriate change to `pauli_to_binary`.

Reply:
I think that makes perfect sense. From my point of view, it comes down to a choice of which functionality could be more intuitive for users. On that note, I'd rather not make a decision solely myself, happy for this to remain as is, or potentially keen to hear thoughts from others too :)